spec.bs

```diff
@@ -2160,6 +2154,12 @@ To <dfn>convert GenerateBidOutput to generated bid</dfn> given a {{GenerateBidOu
   1. Return false.
 </div>
 
+<div algorithm>
+  The <dfn method for="InterestGroupBiddingScriptRunnerGlobalScope">setBid()</dfn> method steps are:
+
+  1. Set [=this=]'s [=relevant global object=]'s [=InterestGroupBiddingScriptRunnerGlobalScope/bid=] to null.
+</div>

```

Comment:
setBid(null) actually does the same in the impl, I am not sure of how to say that nicely.

Reply:
Maybe we should coalesce this together with the other overload, and not have any overloads period. I think https://webidl.spec.whatwg.org/#:~:text=Specification%20authors%20are,not%20using%20overloads applies here, so maybe we should make this `setBid(optional GenerateBidOutput? generateBidOutput)`, what do you think? I think that would make it possible to call:

 - `setBid()`
    - You can detect that the parameter is not given, and set the inner member to null
 - `setBid(undefined)`
    - Treated as the above, per WebIDL, although we'll have to ensure the Chromium implementation matches, since we're not using WebIDL bindings and we might accidentally do something different by default
 - `setBid(null)`
    - Since the dictionary is nullable, you can detect that this is null and act identically to the above
 - `setBid(concreteValue)`
    - Do what the other overload is already doing.

Reply:
I like the idea, but there is a bit of a problem:

[Note: Although dictionary types can in general be nullable, they cannot when used as the type of an operation argument or a dictionary member.](https://webidl.spec.whatwg.org/#idl-nullable-type:~:text=Note%3A%20Although%20dictionary%20types%20can%20in%20general%20be%20nullable%2C%20they%20cannot%20when%20used%20as%20the%20type%20of%20an%20operation%20argument%20or%20a%20dictionary%20member.)

Reply:
I think the restrictions on nullability of dicts in the WebIDL spec are because null auto-converts to an empty dictionary. And we're weird because we accept null as "no bid" but not {}.  So another option is to accept {} as no bid in setBid, rather than have it throw; which spec-wise means making the bid field optional, and explicitly handling its absences.

That's probably not the worst incompat change, mostly means setBid({}) no longer throws, but that's sort of a strange thing to do? (Also changes whether you get a warning about return value to console, which isn't technically an observable change on the web side?)

Reply:
Well, compat-wise doing setBid(null) to say no bid seems to be more reasonable than doing setBid({}) and relying on it throwing.

Reply:
I'm not sure I understand. I guess I'm wondering if we can get rid of the ability to call `setBid(null)` period. I'm proposing that it throw an error because it is invalid, just like `setBid({})` is. Then the way you'd _not_ set a bid (or unset a bid?) via be `setBid()`.

Or if that is not good, then I think it'd be better to just drop the "required" members from the dictionary and use their presence as a signal to set the bid to null vs a real value. I'm not sure which is best, or if there are better ideas. WDYT?

Reply:
I am uncomfortable changing behavior on setBid(null), but a lot more comfortable changing behavior for setBid({}) --- since the latter fails.

So yes, I will likely do the later, once a few other simultaneous things progress enough that I can probably run the corresponding impl through the tests...

Reply:
I think it's done...

Reply:
Catching up on this after a bit, sorry for the delay! OK so just to be clear, the path we've chosen is:
 1. Removing the `required` from the `GenerateBidOutput` dictionary members, so that we can pass both `null` (i.e., implicit `{}`) and explicit `{}` as a bid
 2. Defaulting `bid` to `-1` so that when not included, it goes through the normal null bid path
 3. Detecting when `render` DNE, and returning failure, kicking us back out into the usual failure-handling path

Is that right?

Reply:
Right, though 3 only happens for when bid > 0.